### PR TITLE
Refactor time processing

### PR DESCRIPTION
**Pull request by fsimonis**
## Main changes of this PR

This PR refactors the time processing into a separate class.

* the `DoubleAggregator` is an implementation of a Kahan Babushka Neumaier Aggregator with a bunch of quality of life improvments over our boost variant.
* the `TimeHandler` tracks the current time window

## TODOs

* [x] compute `nextTimeStepSize` in the TimeHandler
* [x] handle `maxTime` in the TimeHandler
* [ ] Add tests for double aggregator 
* [x] Add tests for TimeHandler

## Motivation and additional information

This allows us to test the time handling of very big cases.
This is too expensive to be run as integration tests.

Example:
* `time-window-size=1e-7 max-time=1`
* `time-window-size=0.01 max-time=1000`

## Author's checklist

* [x] I used the [`pre-commit` hook](https://precice.org/dev-docs-dev-tooling.html#setting-up-pre-commit) to prevent dirty commits and used `pre-commit run --all` to format old commits.
* [x] I added a changelog file with `make changelog` if there are user-observable changes since the last release.
* [x] I added a test to cover the proposed changes in our test suite.
* [ ] For breaking changes: I documented the changes in the appropriate [porting guide](https://precice.org/couple-your-code-porting-overview.html).
* [x] I sticked to C++17 features.
* [ ] I sticked to CMake version 3.16.3.
* [ ] I squashed / am about to squash all commits that should be seen as one.

## Reviewers' checklist

* [ ] Does the changelog entry make sense? Is it formatted correctly?
* [ ] Do you understand the code changes?

### CODE REVIEW

File: src/cplscheme/BaseCouplingScheme.cpp [hunk 1]
```diff
@@ -205,14 +211,13 @@ void BaseCouplingScheme::receiveData(const m2n::PtrM2N &m2n, const DataMap &rece
...
   // set _time state to point to end of this window such that _time in receiveData is at end of window
-  _time(_nextTimeWindowSize);
+  _time.progressBy(_nextTimeWindowSize);
   // receive data for end of window
   this->receiveData(m2n, receiveData);
   // reset time state;
-  _time = KahanAccumulator{};
-  _time(oldTime);
+  _time = oldTime;
 }
```
Comment (BenjaminRodenberg) [inline src/cplscheme/BaseCouplingScheme.cpp:221]: This part with `oldTime` still looks strange, but is necessary due to the way how `receiveData` works, if we set `substeps="False"`. Maybe put a comment here that `oldTime` is something temporary. An issue on this topic would be helpful such that I don't have to write this again in every PR review (I know: I should do this and can take care of it).

File: src/cplscheme/BaseCouplingScheme.cpp [hunk 2]
```diff
@@ -433,17 +425,17 @@ bool BaseCouplingScheme::addComputedTime(
...
   PRECICE_CHECK(valid,
                 "The time step size given to preCICE in \"advance\" {} exceeds the maximum allowed time step size {} "
                 "in the remaining of this time window. "
                 "Did you restrict your time step size, \"dt = min(preciceDt, solverDt)\"? "
                 "For more information, consult the adapter example in the preCICE documentation.",
-                timeToAdd, getWindowStartTime() + _timeWindowSize - getTime() + timeToAdd);
+                timeToAdd, getNextTimeWindowSize());
+
+  // add time interval that has been computed in the solver to get the correct time remainder
+  _time.progressBy(timeToAdd);
```
Comment (BenjaminRodenberg) [inline src/cplscheme/BaseCouplingScheme.cpp:438]: This change looks like it is more than a refactoring, but it's definitely a step into the right direction. Are you sure that you might not accidentally introduce something fishy here?

File: src/cplscheme/BaseCouplingScheme.cpp [hunk 3]
```diff
@@ -504,27 +496,20 @@ double BaseCouplingScheme::getNextTimeStepMaxSize() const
...
 
 bool BaseCouplingScheme::isCouplingOngoing() const
 {
-  bool timeLeft      = math::greater(_maxTime, getTime()) || math::equals(_maxTime, UNDEFINED_MAX_TIME);
   bool timestepsLeft = (_maxTimeWindows >= _timeWindows) || (_maxTimeWindows == UNDEFINED_TIME_WINDOWS);
-  return timeLeft && timestepsLeft;
+  bool timeLeft      = !_time.reachedEnd();
+  PRECICE_WARN("tsl:{} tl:{} t:{:e} mt:{:e} umt:{:e} is0:{} isabs0:{}", timestepsLeft, timeLeft, _time.time(), _maxTime, _time.untilTime(_maxTime),
+               math::equals(_time.untilTime(_maxTime), 0.0),
+               math::equals(std::abs(_time.untilTime(_maxTime)), 0.0));
```
Comment (BenjaminRodenberg) [inline src/cplscheme/BaseCouplingScheme.cpp:511]:
```suggestion
```
Debugging statement?

File: src/cplscheme/impl/TimeHandler.hpp
Comment (BenjaminRodenberg) [inline src/cplscheme/impl/TimeHandler.hpp:1]: Did not look into this class in detail, but I guess that's also the good thing about abstraction. I think we need some unit-tests for this class.

File: src/utils/DoubleAggregator.hpp
Comment (BenjaminRodenberg) [inline src/utils/DoubleAggregator.hpp:1]: Why not use the boost version here? I know: The user interface of the boost version is not the best one, but this is something we could hide inside `TimeHandler.hpp`. If we write our own class, there is always the risk of doing something wrong.

**Review by BenjaminRodenberg — CHANGES_REQUESTED** [review]
Generally good approach. Simplifies the (very complicated) `BaseCouplingScheme` and should improve maintainability. We, however, still need some tests (maybe not a must right now).

File: src/cplscheme/BaseCouplingScheme.cpp (hunk 1 above)
Comment (BenjaminRodenberg) [reply src/cplscheme/BaseCouplingScheme.cpp:221]: I tried to summarize the big picture here: https://github.com/precice/precice/issues/1957
@fsimonis would you mind adding a comment with a pointer to https://github.com/precice/precice/issues/1957 in this PR? I hope the issue is somehow understandable. Unfortunately, the situation is confusing. But fixing it needs some effort with low benefit (from the short term perspective).

Comment (fsimonis) [reply src/cplscheme/BaseCouplingScheme.cpp:221]: Done

File: src/cplscheme/BaseCouplingScheme.cpp (hunk 2 above)
Comment (fsimonis) [reply src/cplscheme/BaseCouplingScheme.cpp:438]: Definitely more than a refactoring.

Doesn't this make more sense though?
Check if the step size is permitted and then apply.

File: src/cplscheme/impl/TimeHandler.hpp
Comment (fsimonis) [reply src/cplscheme/impl/TimeHandler.hpp:1]: Added some

File: src/utils/DoubleAggregator.hpp
Comment (fsimonis) [reply src/utils/DoubleAggregator.hpp:1]: Actually, this is so nicely abstracted away that I quickly swapped over to the boost::accumulator.
It is actually more accurate than custom implementation, so I migrated to it.
Thanks for the hint!

**Comment by fsimonis** [comment]
This PR pretty much ready to go but has grown out of a simple refactoring.

The tests that failed in https://github.com/precice/precice/pull/1939#issuecomment-1956588970 do succeed on this branch.

File: src/cplscheme/BaseCouplingScheme.cpp (hunk 2 above)
Comment (BenjaminRodenberg) [reply src/cplscheme/BaseCouplingScheme.cpp:439]: Of course it makes more sense and is simpler to read! I just wanted to say that I do not really know all the implications this might have. But if this part is sufficiently tested: Good to go.

File: src/utils/DoubleAggregator.hpp [hunk 4]
```diff
@@ -0,0 +1,72 @@
+#include <boost/accumulators/accumulators.hpp>
+#include <boost/accumulators/statistics.hpp>
+#include <boost/accumulators/statistics/sum_kahan.hpp>
+
+namespace precice::utils {
+
+/// An accurate aggregator for doubles with usability in mind
+class DoubleAggregator {
+private:
+  using Acc = boost::accumulators::accumulator_set<double, boost::accumulators::stats<boost::accumulators::tag::sum_kahan>>;
```
Comment (BenjaminRodenberg) [inline src/utils/DoubleAggregator.hpp:10]:
```suggestion
  /// This class is mainly a wrapper for boost's Kahan accumulator which implements the Kahan summation algorithm (see https://en.wikipedia.org/wiki/Kahan_summation_algorithm)
  using Acc = boost::accumulators::accumulator_set<double, boost::accumulators::stats<boost::accumulators::tag::sum_kahan>>;
```
I think it would be nice to also mention the keyword "wrapper" here and provide a brief link to what we actually do. Both are implementation details. Therefore, I did not put it into the class description above. Not sure that this is 100% the right approach. Maybe you have a better idea @fsimonis?

File: src/utils/DoubleAggregator.hpp
Comment (BenjaminRodenberg) [inline src/utils/DoubleAggregator.hpp:1]: Looks very nice now!

File: src/cplscheme/impl/TimeHandler.hpp [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+  /// @name Time differences
+  /// @{
+
+  /// Returns the time distance to the possibly truncated end of the current time window
+  double untilWindowEnd(double timeWindowSize) const
+  {
+    double maxDt = untilProgress(timeWindowSize);
+    if (!_maxTime) {
+      return maxDt;
+    }
```
Comment (BenjaminRodenberg) [inline src/cplscheme/impl/TimeHandler.hpp:109]:
```suggestion
```
Is this needed? `untilEnd()` returns `std::numeric_limits<double>::max()` if `!_maxTime`.

File: src/cplscheme/impl/TimeHandler.hpp [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+
+  /** Has the end of the overall time been reached?
+   * This is always false if no max time has been defined.
+   */
+  bool reachedEnd() const
+  {
+    if (!_maxTime) {
+      return false;
+    }
+    return math::equals(untilTime(*_maxTime), 0.0);
```
Comment (BenjaminRodenberg) [inline src/cplscheme/impl/TimeHandler.hpp:96]:
```suggestion
    return math::equals(untilEnd(), 0.0);
```

**Review by BenjaminRodenberg — CHANGES_REQUESTED** [review]
I think there are some minor simplifications. But maybe I'm also just introducing bugs because I don't know it better. Might also be a good motivation for some simple additional tests, if the corner cases I'm touching here are not tested yet.

File: src/utils/DoubleAggregator.hpp (hunk 4 above)
Comment (fsimonis) [reply src/utils/DoubleAggregator.hpp:10]: Good idea.

File: src/cplscheme/BaseCouplingScheme.cpp (hunk 2 above)
Comment (fsimonis) [reply src/cplscheme/BaseCouplingScheme.cpp:439]: The message is actually buggy. :see_no_evil:

File: src/cplscheme/impl/TimeHandler.hpp (hunk 5 above)
Comment (fsimonis) [reply src/cplscheme/impl/TimeHandler.hpp:109]: This is on purpose and should be documented. The reason is two-fold:

1. I try to reduce computations and comparisons of time points as much as possible to avoid accuracy issues. Here we know that it will always be maxDt if there is no maxTime set.
2. This makes the handling of the maxTime easier to spot in the code and makes the flow easier to understand.

File: src/cplscheme/impl/TimeHandler.hpp (hunk 6 above)
Comment (fsimonis) [reply src/cplscheme/impl/TimeHandler.hpp:96]: Same reasoning as above. Instead of risking any issues here, we return early with a clear message. (needs a comment too)

**Review by uekerman — APPROVED** [review]
Looks good to me. Nice job!

**Review by BenjaminRodenberg — APPROVED** [review]
